app/lib/shared/integrity.dart

```diff
@@ -19,296 +19,266 @@ final _logger = Logger('integrity.check');
-  final _packagesWithVersion = <String?>{};
-  final _moderatedPackages = <String?>{};
-  final _publishers = <String?>{};
-  final _publishersAbandoned = <String?>{};
+
+  final _userToOauth = <String, String?>{};
+  final _oauthToUser = <String, String>{};
+  final _deletedUsers = <String>{};
+  final _invalidUsers = <String>{};
+  final _packages = <String>{};
+  final _packageReplacedBys = <String, String>{};
+  final _packagesWithVersion = <String>{};
+  final _publishers = <String>{};
+  final _publishersAbandoned = <String>{};
   int _packageChecked = 0;
   int _versionChecked = 0;
 
   IntegrityChecker(this._db, {int? concurrency})
       : _concurrency = concurrency ?? 1;
 
   /// Runs integrity checks, and reports the list of problems.
-  Future<List<String>> check() async {
-    await _checkUsers();
-    await _checkOAuthUserIDs();
-    await _checkPublishers();
-    await _checkPublisherMembers();
-    await _checkPackages();
-    await _checkVersions();
-    await _checkLikes();
-    await _checkModeratedPackages();
-    return _problems;
+  Stream<String> check() async* {
+    yield* _checkUsers();
+    yield* _checkOAuthUserIDs();
+    yield* _checkPublishers();
+    yield* _checkPublisherMembers();
+    yield* _checkPackages();
+    yield* _checkVersions();
+    yield* _checkLikes();
+    yield* _checkModeratedPackages();
   }
 
-  Future<void> _checkUsers() async {
+  Stream<String> _checkUsers() async* {
     _logger.info('Scanning Users...');
+    final gmailComEmails = <String>{};
     await for (User user in _db.query<User>().run()) {
-      _userToOauth[user.userId] = user.oauthUserId;
-      if (user.email == null ||
-          user.email!.isEmpty ||
-          !looksLikeEmail(user.email)) {
-        _problems.add('User(${user.userId}) has invalid email: ${user.email}');
-        _invalidUsers.add(user.userId);
+      _userToOauth[user.userId!] = user.oauthUserId;
+      final email = user.email;
+      if (email == null || email.isEmpty || !looksLikeEmail(email)) {
+        yield 'User "${user.userId}" has invalid email: "${user.email}".';
+        _invalidUsers.add(user.userId!);
       }
-      if (user.email != null && user.email!.isNotEmpty) {
-        _emailToUser.putIfAbsent(user.email, () => []).add(user.userId);
+
+      // We can have email addresses that have multiple account (also User and
+      // OAuthUserId entities), because at one point they have closed their account,
+      // and somebody (maybe the same person) later re-created one with the given email.
+      //
+      // This is considered normal for non-gmail.com accounts, but should be impossible
+      // for gmail.com accounts, as they cannot be recreated.
+      if (email != null &&
+          email.endsWith('@gmail.com') &&
+          !gmailComEmails.add(email)) {
+        yield 'Email address "$email" is present at "${user.userId}" and another Users.';
       }
+
       if (user.isDeleted is! bool) {
```

Comment:
Hmm with this kind of manual type-checking we could consider having schemas for these entries to automate this...

Not this PR of course

Reply:
Hm, this may be coming from a different era, I think the current Dart SDK would just throw an exception if the value would be non-bool. I think we could just remove these checks for non-nullable values, wdyt?

Reply:
We should probably verify what happens - then I'm all for removing the checks if they are redundant.
If null checks are still relevant we might again want to have a kind of schema-definition for each entity, ensuring that we check all non-nullable fields...

Reply:
Agreed. I think we should do that in a follow-up.
https://github.com/dart-lang/pub-dev/issues/5004